### PR TITLE
add webcam to cameras

### PR DESCRIPTION
**Pull request by ZwX1616**
vipc refactor will happen in a separate PR later

### CODE REVIEW

**Comment by ZwX1616** [comment]
TODO: wait for sensorValid fix, no hacks

File: tools/webcam/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+Run openpilot with webcam on PC/laptop
+=====================
+What's needed:  
+- Ubuntu 16.04  
+- Python 3.7.3  
+- GPU (recommended)  
+- Two USB webcams, at least 720p and 78 degrees FOV (e.g. Logitech C920/C615)  
+- [Car harness](https://comma.ai/shop/products/comma-car-harness) w/ black panda (or the outdated grey panda/giraffe combo) to connect to your car  
+- [Panda paw](https://comma.ai/shop/products/panda-paw) to connect panda to your computer
```
Comment (HaraldSchafer) [inline tools/webcam/README.md]: Is panda paw required?

File: selfdrive/camerad/cameras/camera_webcam.cc [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    cv::Mat frame_mat;
+
+    cap_rear >> frame_mat;
+
+    // int rows = frame_mat.rows;
+    // int cols = frame_mat.cols;
+    // printf("Raw Rear, R=%d, C=%d\n", rows, cols);
+
+    cv::Mat cropped_mat = frame_mat(roi_rear);
+    cv::resize(cropped_mat, frame_mat, cv::Size(s->ci.frame_width, s->ci.frame_height));
```
Comment (HaraldSchafer) [inline selfdrive/camerad/cameras/camera_webcam.cc]: Can you do the proper transform to eon intrinsics? https://github.com/commaai/openpilot/blob/5eeb8a0e02c69bdb1201bf60915a8de51ee74ef8/common/transformations/camera.py#L10

File: selfdrive/modeld/models/commonmodel.c [hunk 3]
```diff
@@ -66,3 +66,11 @@ float sigmoid(float input) {
 float softplus(float input) {
   return log1p(expf(input));
 }
+
+float input_lambda(float input) {
```
Comment (geohot) [inline selfdrive/modeld/models/commonmodel.c]: Is this function being inlined? I'd use a macro here.

Comment (ZwX1616) [reply selfdrive/modeld/models/commonmodel.c]: macro it is

**Comment by HaraldSchafer** [comment]
Looks like you need to bump cereal to get tests to pass?

File: selfdrive/camerad/cameras/camera_webcam.cc [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+static void* rear_thread(void *arg) {
+  int err;
+
+  set_thread_name("webcam_rear_thread");
+  CameraState* s = (CameraState*)arg;
+
+  cv::VideoCapture cap_rear(1); // road
+  cap_rear.set(cv::CAP_PROP_FRAME_WIDTH, 1280);
+  cap_rear.set(cv::CAP_PROP_FRAME_HEIGHT, 720);
+  cap_rear.set(cv::CAP_PROP_FPS, s->fps);
```
Comment (HaraldSchafer) [inline selfdrive/camerad/cameras/camera_webcam.cc]: Does this set the framerate of the camera?

Comment (ZwX1616) [reply selfdrive/camerad/cameras/camera_webcam.cc]: Yes. Found out this can actually limit the model fps

**Comment by geohot** [comment]
Can we merge this before tici camera work?

**Comment by ZwX1616** [comment]
> Can we merge this before tici camera work?

waiting for tests

File: selfdrive/manager.py [hunk 5]
```diff
@@ -204,6 +204,7 @@ def get_running():
   'radard',
   'dmonitoringd',
   'calibrationd',
+  'dmonitoringmodeld',
```
Comment (pd0wm) [inline selfdrive/manager.py]: Only with WEBCAM flag. Make sure it's not added twice. It's also in the ANDROID section.

**Pull request merged**